teuthology/task/install/rpm.py

```diff
@@ -18,49 +17,40 @@ def _remove(ctx, config, remote, rpm):
     :param remote: the teuthology.orchestra.remote.Remote object
     :param rpm: list of packages names to remove
     """
-    rpm = _package_overrides(rpm, remote.os)
+    remote_os = remote.os
+    dist_release = remote_os.name
+    rpm = _package_overrides(rpm, remote_os)
     log.info("Removing packages: {pkglist} on rpm system.".format(
         pkglist=", ".join(rpm)))
-    builder = _get_builder_project(ctx, remote, config)
-    dist_release = builder.dist_release
+    repos = config.get('repos')
 
     if dist_release in ['opensuse', 'sle']:
-        pkg_mng_cmd = 'zypper'
-        pkg_mng_opts = '-n --no-gpg-checks'
-        pkg_mng_subcommand_opts = '--capability'
+        remote.run(args='''
+            for d in {rpms} ; do
+                sudo zypper -n --no-gpg-checks remove --capability $d || true
+            done'''.format(rpms=' '.join(rpm)))
```

Comment:
presumably this will now fail because `||` etc. are not enclosed in run.Raw()

@kshtsk Have you tested this?

Reply:
@smithfarm Yes I tested, check the log: http://158.69.65.8/ubuntu-2017-10-27_15:45:56-deepsea:basic:health-ok-ses5---basic-openstack/65/teuthology.log

> 2017-10-27T16:05:13.922 INFO:teuthology.task.install.rpm:Removing packages: ceph-radosgw, ceph-test, ceph, ceph-mgr, ceph-fuse, libcephfs2, libcephfs-devel, librados2, librbd1, python-ceph, rbd-fuse on rpm system.
2017-10-27T16:05:13.923 INFO:teuthology.orchestra.run.target158069078022:Running: '\n            for d in ceph-radosgw ceph-test ceph ceph-mgr ceph-fuse libcephfs2 libcephfs-devel librados2 librbd1 python-ceph rbd-fuse ; do\n                sudo zypper -n --no-gpg-checks remove --capability $d || true\n            done'
2017-10-27T16:05:14.021 INFO:teuthology.orchestra.run.target158069078022.stdout:Loading repository data...
2017-10-27T16:05:14.021 INFO:teuthology.orchestra.run.target158069078022.stdout:Reading installed packages...
2017-10-27T16:05:14.057 INFO:teuthology.orchestra.run.target158069078022.stdout:Resolving package dependencies...
2017-10-27T16:05:14.063 INFO:teuthology.orchestra.run.target158069078022.stdout:
2017-10-27T16:05:14.063 INFO:teuthology.orchestra.run.target158069078022.stdout:The following package is going to be REMOVED:
2017-10-27T16:05:14.063 INFO:teuthology.orchestra.run.target158069078022.stdout:  ceph-radosgw
2017-10-27T16:05:14.063 INFO:teuthology.orchestra.run.target158069078022.stdout:
2017-10-27T16:05:14.064 INFO:teuthology.orchestra.run.target158069078022.stdout:1 package to remove.
2017-10-27T16:05:14.064 INFO:teuthology.orchestra.run.target158069078022.stdout:After the operation, 11.7 MiB will be freed.
2017-10-27T16:05:14.064 INFO:teuthology.orchestra.run.target158069078022.stdout:Continue? [y/n/...? shows all options] (y): y
2017-10-27T16:05:14.246 INFO:teuthology.orchestra.run.target158069078022.stdout:(1/1) Removing ceph-radosgw-12.2.1-361.gaea79b8b7a.x86_64 [.....done]
2017-10-27T16:05:14.458 INFO:teuthology.orchestra.run.target158069078022.stdout:Loading repository data...
2017-10-27T16:05:14.459 INFO:teuthology.orchestra.run.target158069078022.stdout:Reading installed packages...
2017-10-27T16:05:14.477 INFO:teuthology.orchestra.run.target158069078022.stdout:Resolving package dependencies...
2017-10-27T16:05:14.481 INFO:teuthology.orchestra.run.target158069078022.stdout:
2017-10-27T16:05:14.481 INFO:teuthology.orchestra.run.target158069078022.stdout:The following package is going to be REMOVED:
2017-10-27T16:05:14.481 INFO:teuthology.orchestra.run.target158069078022.stdout:  ceph-test
2017-10-27T16:05:14.482 INFO:teuthology.orchestra.run.target158069078022.stdout:
2017-10-27T16:05:14.482 INFO:teuthology.orchestra.run.target158069078022.stdout:1 package to remove.
2017-10-27T16:05:14.482 INFO:teuthology.orchestra.run.target158069078022.stdout:After the operation, 131.5 MiB will be freed.
2017-10-27T16:05:14.482 INFO:teuthology.orchestra.run.target158069078022.stdout:Continue? [y/n/...? shows all options] (y): y
2017-10-27T16:05:14.542 INFO:teuthology.orchestra.run.target158069078022.stdout:(1/1) Removing ceph-test-12.2.1-361.gaea79b8b7a.x86_64 [.....done]
2017-10-27T16:05:14.719 INFO:teuthology.orchestra.run.target158069078022.stdout:Loading repository data...
2017-10-27T16:05:14.720 INFO:teuthology.orchestra.run.target158069078022.stdout:Reading installed packages...
2017-10-27T16:05:14.738 INFO:teuthology.orchestra.run.target158069078022.stdout:Resolving package dependencies...
2017-10-27T16:05:14.741 INFO:teuthology.orchestra.run.target158069078022.stdout:
2017-10-27T16:05:14.742 INFO:teuthology.orchestra.run.target158069078022.stdout:The following package is going to be REMOVED:
2017-10-27T16:05:14.742 INFO:teuthology.orchestra.run.target158069078022.stdout:  ceph
2017-10-27T16:05:14.742 INFO:teuthology.orchestra.run.target158069078022.stdout:
...
2017-10-27T16:05:17.712 INFO:teuthology.orchestra.run.target158069078022.stdout:
2017-10-27T16:05:17.712 INFO:teuthology.orchestra.run.target158069078022.stdout:Nothing to do.
2017-10-27T16:05:17.784 INFO:teuthology.orchestra.run.target158069078022.stdout:Loading repository data...
2017-10-27T16:05:17.785 INFO:teuthology.orchestra.run.target158069078022.stdout:Reading installed packages...
2017-10-27T16:05:17.807 INFO:teuthology.orchestra.run.target158069078022.stderr:No provider of 'rbd-fuse' found.
2017-10-27T16:05:17.807 INFO:teuthology.orchestra.run.target158069078022.stdout:Resolving package dependencies...
2017-10-27T16:05:17.810 INFO:teuthology.orchestra.run.target158069078022.stdout:
2017-10-27T16:05:17.810 INFO:teuthology.orchestra.run.target158069078022.stdout:Nothing to do.
2017-10-27T16:05:17.818 INFO:teuthology.orchestra.run.target158069078022:Running: 'sudo zypper clean -a'
2017-10-27T16:05:17.961 INFO:teuthology.orchestra.run.target158069078022.stdout:All repositories have been cleaned up.

And it is not looked like it is failing. Why should '||' and ';' failed? The args if it is str is just copied to command which in turn used by exec_command in client module in Paramiko which is passed over channel of ssh session, this is not a command line shell execution.
But I have another question, is there a reason to 'zypper remove' in cycle instead of providing it just a list of rpms? Same question for yum.

Reply:
When the command is run separately for each RPM, and there is a failure, you can tell from the log exactly which RPM it failed on.

Reply:
Yeah, that makes sense, especially if you want to know all failing rpms in one run in contrast to the only first one.

Reply:
Since teuthology aborts at the first failure, you will know only the first one (and you'll know it for certain). Maybe it would be better to put them all together in a single zypper/yum invocation? In the absence of more detailed information, the only worry would be: what if zypper/yum fails with a cryptic error message? Then we don't know which package is problematic.